compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -3679,22 +3702,33 @@ object Types {
   // ----- Annotated and Import types -----------------------------------------------
 
   /** An annotated type tpe @ annot */
-  case class AnnotatedType(tpe: Type, annot: Annotation) extends UncachedProxyType with ValueType {
+  case class AnnotatedType(parent: Type, annot: Annotation) extends UncachedProxyType with ValueType {
     // todo: cache them? but this makes only sense if annotations and trees are also cached.
 
-    override def underlying(implicit ctx: Context): Type = tpe
+    override def underlying(implicit ctx: Context): Type = parent
 
-    def derivedAnnotatedType(tpe: Type, annot: Annotation) =
-      if ((tpe eq this.tpe) && (annot eq this.annot)) this
-      else AnnotatedType(tpe, annot)
+    def derivedAnnotatedType(parent: Type, annot: Annotation) =
+      if ((parent eq this.parent) && (annot eq this.annot)) this
+      else AnnotatedType(parent, annot)
 
     override def stripTypeVar(implicit ctx: Context): Type =
-      derivedAnnotatedType(tpe.stripTypeVar, annot)
+      derivedAnnotatedType(parent.stripTypeVar, annot)
+
+    override def stripAnnots(implicit ctx: Context): Type = parent.stripAnnots
+
+    private[this] var isRefiningKnown = false
+    private[this] var isRefiningCache: Boolean = _
 
-    override def stripAnnots(implicit ctx: Context): Type = tpe.stripAnnots
+    def isRefining(implicit ctx: Context) = {
+      if (!isRefiningKnown) {
```

Comment:
Shouldn't the cache be invalidated at the start of a new run?

Reply:
I was thinking about that too, but decided that it's probably fine given the internal nature of such annotations. Then again, we do expose RefiningAnnotation in the public `scala.annotation` package.